include/envoy/upstream/upstream.h

```diff
@@ -318,6 +318,13 @@ class ClusterInfo {
    *         stats that will be freed when the cluster is removed.
    */
   virtual Stats::Scope& statsScope() const PURE;
+
+  /**
+   * Returns an optional source address for upstream connections to bind to.
+   *
+   * @return an optional source address to bind to.
+   */
+  virtual const Optional<Network::Address::InstanceConstSharedPtr>& sourceAddress() const PURE;
```

Comment:
Fair enough, if that's the least boiler plate then let's do this, in particular as it future proofs.